pkg/provider/keepersecurity/client.go

```diff
@@ -0,0 +1,468 @@
+			return err
+		}
+	} else {
+		if secret.Type() != externalSecretType {
+			return fmt.Errorf(errInvalidSecretType, externalSecretType, secret.Title(), secret.Type())
+		}
+		err = c.updateSecret(secret, parts[1], value)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) DeleteSecret(ctx context.Context, remoteRef esv1beta1.PushRemoteRef) error {
+
+	parts, err := c.buildSecretNameAndKey(remoteRef)
+	if err != nil {
+		return err
+	}
+	secret, err := c.findSecretByName(parts[0])
+	if err != nil {
+		return err
+	}
+	if secret.Type() != externalSecretType {
+		return fmt.Errorf(errInvalidSecretType, externalSecretType, secret.Title(), secret.Type())
+	}
+	_, err = c.ksmClient.DeleteSecrets([]string{secret.Uid})
+	if err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+func (c *Client) buildSecretNameAndKey(remoteRef esv1beta1.PushRemoteRef) ([]string, error) {
+	parts := strings.Split(remoteRef.GetRemoteKey(), "/")
+	if len(parts) != 2 {
+		return nil, fmt.Errorf(errInvalidRemoteRefKey, remoteRef.GetRemoteKey())
+	}
+
+	return parts, nil
+}
+
+func (c *Client) createSecret(name string, key string, value []byte) (string, error) {
+	normalizedKey := strings.ToLower(key)
+	externalSecretRecord := ksm.NewRecordCreate(externalSecretType, name)
+	login := regexp.MustCompile(LoginTypeExpr)
+	pass := regexp.MustCompile(PasswordType)
+	url := regexp.MustCompile(UrlTypeExpr)
+
+	switch {
+	case login.MatchString(normalizedKey):
+		externalSecretRecord.Fields = append(externalSecretRecord.Fields,
+			ksm.NewLogin(string(value)),
+		)
+	case pass.MatchString(normalizedKey):
+		externalSecretRecord.Fields = append(externalSecretRecord.Fields,
+			ksm.NewPassword(string(value)),
+		)
+	case url.MatchString(normalizedKey):
+		externalSecretRecord.Fields = append(externalSecretRecord.Fields,
+			ksm.NewUrl(string(value)),
+		)
+	default:
+		field := ksm.KeeperRecordField{Type: secretType, Label: key}
+		externalSecretRecord.Custom = append(externalSecretRecord.Custom,
+			ksm.Secret{KeeperRecordField: field, Value: []string{string(value)}},
+		)
+	}
+
+	return c.ksmClient.CreateSecretWithRecordData("", c.folderID, externalSecretRecord)
+
+}
+
+func (c *Client) updateSecret(secret *ksm.Record, key string, value []byte) error {
+
+	normalizedKey := strings.ToLower(key)
+	login := regexp.MustCompile(LoginTypeExpr)
+	pass := regexp.MustCompile(PasswordType)
+	url := regexp.MustCompile(UrlTypeExpr)
+
+	switch {
+	case login.MatchString(normalizedKey):
+		secret.SetFieldValueSingle(LoginType, string(value))
+	case pass.MatchString(normalizedKey):
+		secret.SetPassword(string(value))
+	case url.MatchString(normalizedKey):
+		secret.SetFieldValueSingle(UrlType, string(value))
+	default:
+		field := secret.GetCustomFieldValueByLabel(key)
+		if field == "" {
+			return fmt.Errorf(errFieldNotFound, secret.Title(), key)
+		} else {
+			secret.SetCustomFieldValueSingle(key, string(value))
+		}
+	}
+
+	return c.ksmClient.Save(secret)
```

Comment:
I'm a bit concerned with the ksm client implementation. It calls `klog.Panicln` in this function. The code path is triggered when you try to update a secret in a folder that is read-only shared. 

This is a problem - the controller dies because of that. :roll_eyes: 

I think we should open an issue upstream so they remove that behaviour, it's very rude to panic in a sdk :no_entry: 

Also the sdk actively spills out logs :confused:

Reply:
i will open an issue to the core project but they are not really reactive on issues, :p